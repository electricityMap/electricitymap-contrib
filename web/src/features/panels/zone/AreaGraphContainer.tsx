import BreakdownChart from 'features/charts/BreakdownChart';
import CarbonChart from 'features/charts/CarbonChart';
import EmissionChart from 'features/charts/EmissionChart';
import NetExchangeChart from 'features/charts/NetExchangeChart';
import PriceChart from 'features/charts/PriceChart';
import { TimeAverages } from 'utils/constants';

import Divider from './Divider';

export default function AreaGraphContainer({
  datetimes,
  timeAverage,
  displayByEmissions,
  hasEstimationPill,
}: {
  datetimes: Date[];
  timeAverage: TimeAverages;
  displayByEmissions: boolean;
  hasEstimationPill: boolean;
}) {
  return (
    <div>
      {displayByEmissions ? (
        <EmissionChart
          datetimes={datetimes}
          timeAverage={timeAverage}
          hasEstimationPill={hasEstimationPill}
        />
      ) : (
        <CarbonChart
          datetimes={datetimes}
          timeAverage={timeAverage}
          hasEstimationPill={hasEstimationPill}
        />
      )}
      <BreakdownChart
        displayByEmissions={displayByEmissions}
        datetimes={datetimes}
        timeAverage={timeAverage}
        hasEstimationPill={hasEstimationPill}
      />
      <NetExchangeChart datetimes={datetimes} timeAverage={timeAverage} />
      <PriceChart datetimes={datetimes} timeAverage={timeAverage} />
      <Divider />
    </div>
  );
}
