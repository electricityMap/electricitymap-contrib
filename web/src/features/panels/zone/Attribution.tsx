import { GithubButton } from 'components/buttons/GithubButton';
import { useTranslation } from 'react-i18next';

import { getContributors } from './util';

export default function Attribution({ zoneId }: { zoneId: string }) {
  const { t } = useTranslation();
  const { zoneContributorsIndexArray, contributors } = getContributors(zoneId);

  return (
    <div className="flex flex-col gap-3 pt-1.5">
      {zoneContributorsIndexArray.length > 0 && (
        <div className="flex flex-row justify-between">
          <div className="text-sm font-semibold">{t('country-panel.helpfrom')}</div>
          <ContributorList
            zoneContributorsIndexArray={zoneContributorsIndexArray}
            contributors={contributors}
          />
        </div>
      )}
      <GithubButton />
    </div>
  );
}

function ContributorList({
  zoneContributorsIndexArray,
  contributors,
}: {
  zoneContributorsIndexArray: number[];
  contributors: string[];
}) {
  if (!zoneContributorsIndexArray) {
    return null;
  }

  return (
    <div className="flex flex-wrap gap-1">
      {zoneContributorsIndexArray.map((contributorIndex) => (
        <a
          key={contributors.at(contributorIndex)}
          href={`https://github.com/${contributors.at(contributorIndex)}`}
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            src={`https://avatars.githubusercontent.com/${contributors.at(
              contributorIndex
            )}?s=20`} // loads the avatar image at a default size of 20px
            srcSet={`https://avatars.githubusercontent.com/${contributors.at(
              contributorIndex
            )}?s=40 2x`} // loads the avatar image at a default size of 40px for high resolution displays
            alt={contributors.at(contributorIndex)}
            height="20"
            width="20"
            loading="lazy" // makes sure the image don't load until the user scrolls down
            className="rounded-full"
          />
        </a>
      ))}
    </div>
  );
}
