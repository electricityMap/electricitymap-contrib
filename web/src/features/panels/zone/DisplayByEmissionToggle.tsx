import ToggleButton from 'components/ToggleButton';
import { useAtom, useAtomValue } from 'jotai';
import type { ReactElement } from 'react';
import trackEvent from 'utils/analytics';
import { LeftPanelToggleOptions } from 'utils/constants';
import { displayByEmissionsAtom, isConsumptionAtom } from 'utils/state/atoms';

export default function EmissionToggle(): ReactElement {
  const isConsumption = useAtomValue(isConsumptionAtom);
  const [displayByEmissions, setDisplayByEmissions] = useAtom(displayByEmissionsAtom);

  // TODO: perhaps togglebutton should accept boolean values
  const options = [
    {
      value: LeftPanelToggleOptions.ELECTRICITY,
      translationKey: isConsumption
        ? 'country-panel.electricityconsumption'
        : 'country-panel.electricityproduction',
    },
    {
      value: LeftPanelToggleOptions.EMISSIONS,
      translationKey: 'country-panel.emissions',
    },
  ];

  const onSetCurrentMode = (option: string) => {
    if (displayByEmissions) {
      trackEvent('PanelProductionButton Clicked');
    } else {
      trackEvent('PanelEmissionButton Clicked');
    }
    if (
      (option === LeftPanelToggleOptions.ELECTRICITY && displayByEmissions) ||
      (option === LeftPanelToggleOptions.EMISSIONS && !displayByEmissions)
    ) {
      setDisplayByEmissions(!displayByEmissions);
    }
  };

  return (
    <div className="my-4">
      <ToggleButton
        options={options}
        selectedOption={
          displayByEmissions
            ? LeftPanelToggleOptions.EMISSIONS
            : LeftPanelToggleOptions.ELECTRICITY
        }
        onToggle={onSetCurrentMode}
      />
    </div>
  );
}
