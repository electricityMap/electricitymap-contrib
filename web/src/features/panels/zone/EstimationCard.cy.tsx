import EstimationCard from './EstimationCard';

describe('EstimationCard', () => {
  beforeEach(() => {
    cy.mount(
      <EstimationCard
        cardType="estimated"
        estimationMethod="ESTIMATED_CONSTRUCT_BREAKDOWN"
        outageMessage={undefined}
      />
    );
  });

  it('feedback card should only be visible when collapse button has been clicked twice or more', () => {
    cy.get('[data-test-id=feedback-card]').should('not.exist');
    cy.get('[data-test-id=collapse-button]').click();
    cy.get('[data-test-id=feedback-card]').should('exist');
    cy.get('[data-test-id=collapse-button]').click();
    cy.get('[data-test-id=feedback-card]').should('exist');
  });
});
