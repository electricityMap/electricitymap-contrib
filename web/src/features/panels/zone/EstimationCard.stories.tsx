import type { Meta, StoryObj } from '@storybook/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { I18nextProvider } from 'react-i18next';
import i18n from 'translation/i18n';

import EstimationCard from './EstimationCard';

const meta: Meta<typeof EstimationCard> = {
  title: 'EstimationCard',
  component: EstimationCard,
};

export default meta;
type Story = StoryObj<typeof EstimationCard>;
const instance: { message: string; issue: string } = {
  message: 'Some outage message.',
  issue: '5613',
};
const queryClient = new QueryClient();

export const All: Story = {
  render: () => (
    <div className="space-y-2">
      <I18nextProvider i18n={i18n}>
        <QueryClientProvider client={queryClient}>
          <EstimationCard cardType="outage" zoneMessage={instance} />
          <EstimationCard
            cardType="outage"
            zoneMessage={instance}
            estimationMethod="threshold_filtered"
          />
          <EstimationCard cardType="estimated" zoneMessage={instance} />

          <EstimationCard
            cardType="estimated"
            zoneMessage={instance}
            estimationMethod="estimated_mode_breakdown"
          />

          <EstimationCard
            cardType="aggregated"
            zoneMessage={instance}
            estimatedPercentage={20}
          />

          <EstimationCard cardType="aggregated" zoneMessage={instance} />
        </QueryClientProvider>
      </I18nextProvider>
    </div>
  ),
};
