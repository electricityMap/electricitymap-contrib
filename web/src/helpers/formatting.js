/* eslint-disable */
// TODO: remove once refactored

import * as d3 from 'd3-format';
import { TIME } from './constants';
import * as translation from './translation';

const formatPower = function (d, numDigits) {
  // Assume MW input
  if (d == null || d === NaN) return d;
  if (numDigits == null) numDigits = 3;
  return d3.format('.' + numDigits + 's')(d * 1e6) + 'W';
};
const formatCo2 = function (d, numDigits) {
  // Assume gCO₂ / h input
  d /= 60; // Convert to gCO₂ / min
  d /= 1e6; // Convert to tCO₂ / min
  if (d == null || d === NaN) return d;
  if (numDigits == null) numDigits = 3;
  if (d >= 1)
    // a ton or more
    return d3.format('.' + numDigits + 's')(d) + 't ' + translation.translate('ofCO2eqPerMinute');
  else return d3.format('.' + numDigits + 's')(d * 1e6) + 'g ' + translation.translate('ofCO2eqPerMinute');
};
const scalePower = function (maxPower) {
  // Assume MW input
  if (maxPower < 1)
    return {
      unit: 'kW',
      formattingFactor: 1e-3,
    };
  if (maxPower < 1e3)
    return {
      unit: 'MW',
      formattingFactor: 1,
    };
  else
    return {
      unit: 'GW',
      formattingFactor: 1e3,
    };
};

const formatDate = function (date, lang, time) {
  if (!date || !time) return '';

  switch (time) {
    case TIME.HOURLY:
      return new Intl.DateTimeFormat(lang, { dateStyle: 'long', timeStyle: 'short' }).format(date);
    case TIME.DAILY:
      return new Intl.DateTimeFormat(lang, { dateStyle: 'long' }).format(date);
    default:
      console.error(`${time} is not implemented`);
      return '';
  }
};

const getLocaleUnit = (dateUnit, lang) =>
  new Intl.NumberFormat(lang, {
    style: 'unit',
    unit: dateUnit,
    unitDisplay: 'long',
  })
    .formatToParts(1)
    .filter((x) => x.type === 'unit')[0].value;

const getLocaleNumberFormat = (lang, { unit, unitDisplay, range }) =>
  new Intl.NumberFormat(lang, {
    style: 'unit',
    unit: unit,
    unitDisplay: unitDisplay || 'long',
  }).format(range);

const formatTimeRange = (lang, timeAggregate) => {
  // Note that not all browsers fully support all languages
  switch (timeAggregate) {
    case TIME.HOURLY:
      return getLocaleNumberFormat(lang, { unit: 'hour', range: 24 });
    case TIME.DAILY:
      return getLocaleUnit('month', lang);
    case TIME.MONTHLY:
      return getLocaleUnit('year', lang);
    case TIME.YEARLY:
      return getLocaleNumberFormat(lang, { unit: 'year', range: 5 });
    default:
      console.error(`${time} is not implemented`);
      return '';
  }
};

const formatDateTick = function (date, lang, time) {
  if (!date || !time) return '';

  switch (time) {
    case TIME.HOURLY:
      return new Intl.DateTimeFormat(lang, { timeStyle: 'short' }).format(date);
    case TIME.DAILY:
      return new Intl.DateTimeFormat(lang, { month: 'long', day: 'numeric' }).format(date);
    default:
      console.error(`${time} is not implemented`);
      return '';
  }
};

export { formatPower, formatCo2, scalePower, formatDate, formatTimeRange, formatDateTick };
