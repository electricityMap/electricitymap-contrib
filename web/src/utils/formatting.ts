import * as d3 from 'd3-format';
import { TimeAverages } from './constants';
import { EnergyUnits } from './units';

const DEFAULT_NUM_DIGITS = 2;

export const formatPower = function (
  d: number,
  numberDigits: number = DEFAULT_NUM_DIGITS
) {
  // Assume MW input
  if (d == undefined || Number.isNaN(d)) {
    return d;
  }
  const power = `${d3.format(`.${numberDigits}s`)(d * 1e6)}Wh` //Add a space between the number and the unit
    .replace(/([A-Za-z])/, ' $1')
    .trim();
  return power;
};

export const formatCo2 = function (gramPerHour: number, valueToMatch?: number) {
  if (gramPerHour == undefined || Number.isNaN(gramPerHour)) {
    return gramPerHour;
  }

  // Assume gCO₂ / h input
  let value = gramPerHour;
  value /= 60; // Convert to gCO₂ / min
  value /= 1e6; // Convert to tCO₂ / min

  // Ensure both numbers are at the same scale
  const checkAgainst = valueToMatch ? valueToMatch / 1e6 : value;

  // grams and kilograms
  if (checkAgainst < 1) {
    return `${d3.format(`,.0~s`)(value * 1e6)}g`;
  }

  // tons
  if (Math.round(checkAgainst) < 1e5) {
    const decimals = value < 1 ? 2 : 1;
    return `${d3.format(`,.${decimals}~f`)(value)}t`;
  }

  // Hundred thousands of tons
  if (Math.round(checkAgainst) < 1e6) {
    return `${d3.format(`,.1~f`)(value / 1e6)}Mt`;
  }

  // megatons or above
  return `${d3.format(`,.2~s`)(value)}t`;
};

const scalePower = function (maxPower: number | undefined) {
  // Assume MW input
  if (maxPower == undefined) {
    return {
      unit: '?',
      formattingFactor: 1e3,
    };
  }

  const thresholds: [number, EnergyUnits][] = [
    [1e9, EnergyUnits.PETAWATT_HOURS],
    [1e6, EnergyUnits.TERAWATT_HOURS],
    [1e3, EnergyUnits.GIGAWATT_HOURS],
    [1, EnergyUnits.MEGAWATT_HOURS],
    [1e-3, EnergyUnits.KILOWATT_HOURS],
  ];

  // Use absolute value to handle negative values
  const value = Math.abs(maxPower);

  for (const [threshold, unit] of thresholds) {
    if (value >= threshold) {
      return {
        unit,
        formattingFactor: threshold,
      };
    }
  }

  // Fallback if none of the thresholds are met
  return {
    unit: EnergyUnits.PETAWATT_HOURS,
    formattingFactor: 1e9,
  };
};

const formatDate = function (date: Date, lang: string, time: string) {
  if (!isValidDate(date) || !time) {
    return '';
  }

  switch (time) {
    case TimeAverages.HOURLY: {
      return new Intl.DateTimeFormat(lang, {
        dateStyle: 'medium',
        timeStyle: 'short',
      }).format(date);
    }
    // Instantiate below DateTimeFormat objects using UTC to avoid displaying
    // misleading time slider labels for users in UTC-negative offset timezones
    case TimeAverages.DAILY: {
      return new Intl.DateTimeFormat(lang, {
        dateStyle: 'long',
        timeZone: 'UTC',
      }).format(date);
    }
    case TimeAverages.MONTHLY: {
      return new Intl.DateTimeFormat(lang, {
        month: 'long',
        year: 'numeric',
        timeZone: 'UTC',
      }).format(date);
    }
    case TimeAverages.YEARLY: {
      return new Intl.DateTimeFormat(lang, {
        year: 'numeric',
        timeZone: 'UTC',
      }).format(date);
    }
    default: {
      console.error(`${time} is not implemented`);
      return '';
    }
  }
};

const getLocaleNumberFormat = (lang: string, { unit, unitDisplay, range }: any) => {
  try {
    return new Intl.NumberFormat(lang, {
      style: 'unit',
      unit,
      unitDisplay: unitDisplay || 'long',
    }).format(range);
  } catch {
    // As Intl.NumberFormat with custom 'unit' is not supported in all browsers, we fallback to
    // a simple English based implementation
    const plural = range === 1 ? '' : 's';
    return `${range} ${unit}${plural}`;
  }
};

const formatTimeRange = (lang: string, timeAggregate: TimeAverages) => {
  // Note that not all browsers fully support all languages
  switch (timeAggregate) {
    case TimeAverages.HOURLY: {
      return getLocaleNumberFormat(lang, { unit: 'hour', range: 24 });
    }
    case TimeAverages.DAILY: {
      return getLocaleNumberFormat(lang, { unit: 'day', range: 30 });
    }
    case TimeAverages.MONTHLY: {
      return getLocaleNumberFormat(lang, { unit: 'month', range: 12 });
    }
    case TimeAverages.YEARLY: {
      return getLocaleNumberFormat(lang, {
        unit: 'year',
        range: new Date().getUTCFullYear() - 2017,
      });
    }
    default: {
      console.error(`${timeAggregate} is not implemented`);
      return '';
    }
  }
};

const formatDateTick = function (date: Date, lang: string, timeAggregate: TimeAverages) {
  if (!isValidDate(date) || !timeAggregate) {
    return '';
  }

  switch (timeAggregate) {
    case TimeAverages.HOURLY: {
      return new Intl.DateTimeFormat(lang, {
        timeStyle: 'short',
      }).format(date);
    }
    // Instantiate below DateTimeFormat objects using UTC to avoid displaying
    // misleading time slider labels for users in UTC-negative offset timezones
    case TimeAverages.DAILY: {
      return new Intl.DateTimeFormat(lang, {
        month: 'short',
        day: 'numeric',
        timeZone: 'UTC',
      }).format(date);
    }
    case TimeAverages.MONTHLY: {
      return lang === 'et'
        ? new Intl.DateTimeFormat(lang, {
            month: 'short',
            day: 'numeric',
            timeZone: 'UTC',
          })
            .formatToParts(date)
            .find((part) => part.type === 'month')?.value
        : new Intl.DateTimeFormat(lang, {
            month: 'short',
            timeZone: 'UTC',
          }).format(date);
    }
    case TimeAverages.YEARLY: {
      return new Intl.DateTimeFormat(lang, {
        year: 'numeric',
        timeZone: 'UTC',
      }).format(date);
    }
    default: {
      console.error(`${timeAggregate} is not implemented`);
      return '';
    }
  }
};

function isValidDate(date: Date) {
  if (!date || !(date instanceof Date)) {
    return false;
  }

  if (!date?.getTime() || date?.getTime() <= 1) {
    return false;
  }

  return true;
}
/**
 * @param {string[]} dataSources - array of data sources.
 * @param {string} language - ISO 639-1 language code (`en`) or ISO 639-1 language code + ISO 3166-1 alpha-2 country code (`en-GB`).
 * @returns {string} formatted string of data sources.
 */
function formatDataSources(dataSources: string[], language: string) {
  return Intl.ListFormat === undefined
    ? dataSources.join(', ')
    : new Intl.ListFormat(language, { style: 'long', type: 'conjunction' }).format(
        dataSources
      );
}

export { formatDataSources, formatDate, formatDateTick, formatTimeRange, scalePower };
