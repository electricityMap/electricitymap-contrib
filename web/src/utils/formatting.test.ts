import { removeDuplicateSources } from 'features/panels/zone/Attribution';

import { formatCo2, formatDataSources, formatEnergy, formatPower } from './formatting';

describe('formatEnergy', () => {
  it('handles NaN input', () => {
    const actual = formatEnergy(Number.NaN);
    const expected = Number.NaN;
    expect(actual).toBe(expected);
  });

  it('handles custom number of digits', () => {
    const actual = formatEnergy(1.234_567, 4);
    const expected = '1.235 MWh';
    expect(actual).toBe(expected);
  });

  it('handles default number kWh', () => {
    const actual = formatEnergy(0.002_234_567);
    const expected = '2.2 kWh';
    expect(actual).toBe(expected);
  });

  it('handles MWh', () => {
    const actual = formatEnergy(1.234_567);
    const expected = '1.2 MWh';
    expect(actual).toBe(expected);
  });

  it('handles GWh', () => {
    const actual = formatEnergy(1222.234_567);
    const expected = '1.2 GWh';
    expect(actual).toBe(expected);
  });

  it('handles TWh', () => {
    const actual = formatEnergy(1_222_000.234_567);
    const expected = '1.2 TWh';
    expect(actual).toBe(expected);
  });

  it('Converts PWh to TWh', () => {
    const actual = formatEnergy(1_222_000_000.234_567);
    const expected = '1200 TWh';
    expect(actual).toBe(expected);
  });

  it('handles zero input', () => {
    const actual = formatEnergy(0);
    const expected = '0 Wh';
    expect(actual).toBe(expected);
  });

  it('handles 1 input for number of digits', () => {
    const actual = formatEnergy(12_313, 1);
    const expected = '10 GWh';
    expect(actual).toBe(expected);
  });

  it('handles 0 input for number of digits', () => {
    const actual = formatEnergy(12_313, 0);
    const expected = '10 GWh';
    expect(actual).toBe(expected);
  });
});

describe('formatPower', () => {
  it('handles NaN input', () => {
    const actual = formatPower(Number.NaN);
    const expected = Number.NaN;
    expect(actual).toBe(expected);
  });

  it('handles custom number of digits', () => {
    const actual = formatPower(1.234_567, 4);
    const expected = '1.235 MW';
    expect(actual).toBe(expected);
  });

  it('handles default number kW', () => {
    const actual = formatPower(0.002_234_567);
    const expected = '2.2 kW';
    expect(actual).toBe(expected);
  });

  it('handles MW', () => {
    const actual = formatPower(1.234_567);
    const expected = '1.2 MW';
    expect(actual).toBe(expected);
  });

  it('handles GW', () => {
    const actual = formatPower(1222.234_567);
    const expected = '1.2 GW';
    expect(actual).toBe(expected);
  });

  it('handles TW', () => {
    const actual = formatPower(1_222_000.234_567);
    const expected = '1.2 TW';
    expect(actual).toBe(expected);
  });

  it('handles zero input', () => {
    const actual = formatPower(0);
    const expected = '0 W';
    expect(actual).toBe(expected);
  });

  it('handles 1 input for number of digits', () => {
    const actual = formatPower(12_313, 1);
    const expected = '10 GW';
    expect(actual).toBe(expected);
  });

  it('handles 0 input for number of digits', () => {
    const actual = formatPower(12_313, 0);
    const expected = '10 GW';
    expect(actual).toBe(expected);
  });
});

describe('formatCo2', () => {
  it('handles grams', () => {
    const actual = formatCo2(20);
    const expected = '20 g';
    expect(actual).toBe(expected);
  });
  it('handles kilograms', () => {
    const actual = formatCo2(1000);
    const expected = '1 kg';
    expect(actual).toBe(expected);
  });
  it('handles tonnes', () => {
    const actual = formatCo2(1_000_000);
    const expected = '1 t';
    expect(actual).toBe(expected);
  });

  it('uses same unit as another value would', () => {
    const actual = formatCo2(23_500, 2_350_000);
    const expected = '0.02 t';
    expect(actual).toBe(expected);
  });

  it('adds decimals if comparing with tonnes', () => {
    const actual = formatCo2(200_500, 2_350_000);
    const expected = '0.2 t';
    expect(actual).toBe(expected);
  });
  it('adds decimals if comparing with large tonnes', () => {
    const actual = formatCo2(200_500, 992_350_000);
    const expected = '0.2 t';
    expect(actual).toBe(expected);
  });
  it('handles real data value', () => {
    const actual = formatCo2(740_703_650);
    const expected = '740 t';
    expect(actual).toBe(expected);
  });
  it('handles kilotonnes', () => {
    const actual = formatCo2(99_000_000_000);
    const expected = '99 kt';
    expect(actual).toBe(expected);
  });
  it('handles megatonnes', () => {
    const actual = formatCo2(99_000_000_000_000);
    const expected = '99 Mt';
    expect(actual).toBe(expected);
  });
  it('handles megatonnes close to 1Gt rounding down', () => {
    const actual = formatCo2(994_320_320_231_123);
    const expected = '990 Mt';
    expect(actual).toBe(expected);
  });
  it('handles values up to 1 trillion grams, rounding up', () => {
    const actual = formatCo2(999_000_000_000_000);
    const expected = '1 Gt';
    expect(actual).toBe(expected);
  });
  it('handles values above 1 trillion', () => {
    const actual = formatCo2(6_700_000_000_000_000);
    const expected = '6.7 Gt';
    expect(actual).toBe(expected);
  });
  it('handles values petatonnes', () => {
    const actual = formatCo2(1.5e21);
    const expected = '1.5 Pt';
    expect(actual).toBe(expected);
  });
});

describe('formatDataSources', () => {
  it('handles multiple sources with en', () => {
    const input = removeDuplicateSources(
      `"electricityMap Estimation","entsoe.eu","entsoe.eu"`
    );

    const expected = 'electricityMap Estimation and entsoe.eu';

    const actual = formatDataSources(input, 'en');
    expect(actual).toBe(expected);
  });

  it('handles multiple sources in another language', () => {
    const input = removeDuplicateSources(
      `"electricityMap Estimation","entsoe.eu","entsoe.eu"`
    );

    const expected = 'electricityMap Estimation og entsoe.eu';

    const actual = formatDataSources(input, 'da');
    expect(actual).toBe(expected);
  });

  it('single source is not changed', () => {
    const input = removeDuplicateSources(`"entsoe.eu"`);

    const expected = 'entsoe.eu';

    const actual = formatDataSources(input, 'en');
    expect(actual).toBe(expected);
  });
});
