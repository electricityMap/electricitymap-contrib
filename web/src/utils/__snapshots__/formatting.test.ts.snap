// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatDate > handles daily data for de 1`] = `"1.–2. Januar 2021"`;

exports[`formatDate > handles daily data for en 1`] = `"January 1 – 2, 2021"`;

exports[`formatDate > handles daily data for es 1`] = `"1–2 de enero de 2021"`;

exports[`formatDate > handles daily data for fr 1`] = `"1–2 janvier 2021"`;

exports[`formatDate > handles daily data for it 1`] = `"01–02 gennaio 2021"`;

exports[`formatDate > handles daily data for sv 1`] = `"1–2 januari 2021"`;

exports[`formatDate > handles hourly data for de 1`] = `"1. Jan. 2021, 00:00–01:00 Uhr UTC"`;

exports[`formatDate > handles hourly data for en 1`] = `"Jan 1, 2021, 12:00 – 1:00 AM UTC"`;

exports[`formatDate > handles hourly data for es 1`] = `"1 ene 2021, 0:00–1:00 UTC"`;

exports[`formatDate > handles hourly data for fr 1`] = `"1 janv. 2021, 00:00 – 01:00 UTC"`;

exports[`formatDate > handles hourly data for it 1`] = `"1 gen 2021, 00:00–01:00 UTC"`;

exports[`formatDate > handles hourly data for sv 1`] = `"1 jan. 2021 00:00–01:00 UTC"`;

exports[`formatDate > handles monthly data for de 1`] = `"Januar–Februar 2021"`;

exports[`formatDate > handles monthly data for en 1`] = `"January – February 2021"`;

exports[`formatDate > handles monthly data for es 1`] = `"enero–febrero de 2021"`;

exports[`formatDate > handles monthly data for fr 1`] = `"janvier – février 2021"`;

exports[`formatDate > handles monthly data for it 1`] = `"gennaio–febbraio 2021"`;

exports[`formatDate > handles monthly data for sv 1`] = `"januari–februari 2021"`;

exports[`formatDate > handles yearly data for de 1`] = `"2021–2022"`;

exports[`formatDate > handles yearly data for en 1`] = `"2021 – 2022"`;

exports[`formatDate > handles yearly data for es 1`] = `"2021–2022"`;

exports[`formatDate > handles yearly data for fr 1`] = `"2021–2022"`;

exports[`formatDate > handles yearly data for it 1`] = `"2021–2022"`;

exports[`formatDate > handles yearly data for sv 1`] = `"2021–2022"`;
